chalk-ir/src/lib.rs

```diff
@@ -1013,6 +1010,52 @@ impl<T> UCanonical<T> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasTypeFamily)]
+/// A general goal; this is the full range of questions you can pose to Chalk.
+pub struct Goals<TF: TypeFamily> {
+    goals: Vec<Goal<TF>>,
+}
+
+impl<TF: TypeFamily> Goals<TF> {
+    pub fn from(goals: impl IntoIterator<Item = impl CastTo<Goal<TF>>>) -> Self {
+        use crate::cast::Caster;
+        Goals {
+            goals: goals.into_iter().casted().collect(),
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::from(None::<Goal<TF>>)
```

Comment:
This works, but I think I'd prefer the straightforward `Goals { goals: vec![] }`

Reply:
I didn't do that because I intend to refactor this more later and make it interned. I should maybe do that in this same PR though, forgot I took the first steps.

Reply:
Did the rest of the refactoring now.